### ISSUE TITLE
Issue with model setup

### ISSUE THREAD

**Issue by AbelThar** [adriangb/scikeras#188]
I am having issues getting SciKeras getting to work as expected; 

From evaluating the result it seems that SciKeras learns as expected (at least shown in verbose output) but does not modify the model (as predictions made by this model are all wrong).

I used this same model directly with Keras and got a reasonable score.

I am not sure if this is an installation issue (running this on Conda through pip v0.2.1), or if I overlooked a parameter while setting it up for my use case.
I know I left this a bit dry, I would appreciate some help knowing what to look for, and let me know what I can provide.

### Imports

```python
# import rfpimp
import pickle
import dask.dataframe as dd

# Custom Implementation
from modules.classification import Classification, ConfusionMatrix, RFVisualisation
```

```python
from tensorflow import keras
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import train_test_split, cross_val_score
import numpy as np
```

### Running Quickstart Example seems to work as expected...

```python
import numpy as np
from sklearn.datasets import make_classification
from tensorflow import keras

from scikeras.wrappers import KerasClassifier


X, y = make_classification(1000, 20, n_informative=10, random_state=0)
X = X.astype(np.float32)
y = y.astype(np.int64)

def get_model(hidden_layer_dim, meta):
    # note that meta is a special argument that will be
    # handed a dict containing input metadata
    n_features_in_ = meta["n_features_in_"]
    X_shape_ = meta["X_shape_"]
    n_classes_ = meta["n_classes_"]

    model = keras.models.Sequential()
    model.add(keras.layers.Dense(n_features_in_, input_shape=X_shape_[1:]))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(hidden_layer_dim))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(n_classes_))
    model.add(keras.layers.Activation("softmax"))
    return model

clf = KerasClassifier(
    get_model,
    loss="sparse_categorical_crossentropy",
    hidden_layer_dim=100,
)

clf.fit(X, y)
y_proba = clf.predict_proba(X)
y_proba
```

    32/32 [==============================] - 1s 1ms/step - loss: 0.6739
    32/32 [==============================] - 0s 756us/step

    array([[0.4033311 , 0.59666884],
           [0.78341705, 0.21658295],
           [0.5785239 , 0.4214761 ],
           ...,
           [0.8525218 , 0.14747825],
           [0.5348476 , 0.46515232],
           [0.63776696, 0.36223304]], dtype=float32)

### Will run with my existing implementation...

```python
cl = Classification(verbose=True)

def evaluate_sample(featureset, description, classifier=default_classifier, 
                    sample=path['samples'], scaler=StandardScaler, 
                    validation=0.2, imbalance=None,
                    allcolumns=allcolumns, label='code', tileset='CVieL', 
                    nomenclature=nomenclature, resultspath=path['results'], 
                    verbose=True, overwrite=False):
    
    cl.vprint(f'Description:\t{description}')
    
    pkl_path = Path(resultspath, 'pickles', f'{description}.pickle')
    if not overwrite and pkl_path.is_file():
        cl.vprint('Run Exists! Loading from pickle...')
        return load_result_pickle(pkl_path)        
    
    cols = ['code']+[c for l in featureset for c in allcolumns if c.endswith(l)]
    featuresdf = cl.load_features(sample, cols)
    
    if imbalance is not None:
        featuresdf = cl.resample_features(featuresdf, imbalance)
    else:
        szdf = featuresdf.groupby(by='code').size()
        imbalance = round(szdf.max()/szdf.min(), 1)
            
    splits, featcols, validation = cl.split_features(featuresdf, label, validation, scaler)
    X_train, X_test, y_train, y_test = splits
    
    return splits

X_train, X_test, y_train, y_test = evaluate_sample(**{'featureset': o_ind_stats, 'description': 'S2 indices mean and variance with SciKeras'})
```

    Description:	S2 indices mean and variance with SciKeras
    Split Ratio:	0.2
    Split Shapes:	[(79997, 96), (20000, 96), (79997, 12), (20000, 12)]

### Using SciKeras, with failing scores

```python
# TODO: Fix SciKeras for scikit-learn integration
def fcn_model(hidden_layer_dim, meta):
    
    print('meta: {\n', '\n  '.join([f'{kv}' for kv in meta.items()]), '\n}')
    # note that meta is a special argument that will be
    # handed a dict containing input metadata
    n_features_in_ = meta["n_features_in_"]
    X_shape_ = meta["X_shape_"]
    n_classes_ = meta["n_classes_"]
    
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(n_features_in_, input_shape=X_shape_[1:]))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(hidden_layer_dim))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(n_classes_))
    model.add(keras.layers.Activation("softmax"))
    return model

def get_model_fcn(modelargs={}):
    return KerasClassifier(fcn_model, 
                           hidden_layer_dim=128, 
                           epochs=1,
                           optimizer='adam',
                           loss='categorical_crossentropy',
                           metrics=['accuracy'],
                           **modelargs)

model = get_model_fcn()
model.fit(X_train, y_train)

columns, report, true_matrix, pred_matrix = cl.classification_metrics(model, [X_train, X_test, y_train, y_test], [0, 0])
```

    meta: {
     ('classes_', array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]))
      ('target_type_', 'multilabel-indicator')
      ('y_dtype_', dtype('uint8'))
      ('y_ndim_', 2)
      ('X_dtype_', dtype('float32'))
      ('X_shape_', (79997, 96))
      ('n_features_in_', 96)
      ('target_encoder_', ClassifierLabelEncoder(loss='categorical_crossentropy'))
      ('n_classes_', 12)
      ('n_outputs_', 1)
      ('n_outputs_expected_', 1)
      ('feature_encoder_', FunctionTransformer()) 
    }
    2500/2500 [==============================] - 3s 1ms/step - loss: 1.2360 - accuracy: 0.5530
    625/625 [==============================] - 0s 686us/step
    Overall: 0.130  Kappa: 0.000
    Weighted F1-Score: 0.030

    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))
    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))
    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))

### Using same SciKeras model, directly with Keras, is working

```python
# Co-Opted from SciKeras Model Compilation
meta = {
    'classes_': np.array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]), 
    'target_type_': 'multilabel-indicator', 
    'y_dtype_': np.dtype('uint8'), 
    'y_ndim_': 2, 
    'X_dtype_': np.dtype('float32'), 
    'X_shape_': (79997, 96), 
    'n_features_in_': 96, 
    #'target_encoder_': ClassifierLabelEncoder(loss='categorical_crossentropy'), 
    'n_classes_': 12, 
    'n_outputs_': 1, 
    'n_outputs_expected_': 1, 
    #'feature_encoder_': FunctionTransformer()
}

model = fcn_model(128, meta)
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=1)
columns, report, true_matrix, pred_matrix = cl.classification_metrics(model, [X_train, X_test, y_train, y_test], [0, 0])
```

    meta: {
     ('classes_', array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]))
      ('target_type_', 'multilabel-indicator')
      ('y_dtype_', dtype('uint8'))
      ('y_ndim_', 2)
      ('X_dtype_', dtype('float32'))
      ('X_shape_', (79997, 96))
      ('n_features_in_', 96)
      ('n_classes_', 12)
      ('n_outputs_', 1)
      ('n_outputs_expected_', 1) 
    }
    2500/2500 [==============================] - 3s 1ms/step - loss: 1.2370 - accuracy: 0.5477
    Overall: 0.660  Kappa: 0.615
    Weighted F1-Score: 0.651

### Classification Metrics Implementation

```python
# Model Analysis Metrics
def classification_metrics(self, model, splits, score):
    X_train, X_test, y_train, y_test = splits
    columns, labels, predictions = y_train.columns, y_test.values, model.predict(X_test)

    # Generate Report using Test Set predictions
    labels_list = labels.argmax(axis=1)
    predictions_list = predictions.argmax(axis=1)

    report = classification_report(labels_list, predictions_list, 
                                   target_names=columns, output_dict=True)

    # Add Train Set imbalance to the report
    imbalance = y_train.sum(axis=0)
    imbalance = imbalance/min(imbalance)

    for code, code_imbalance in imbalance.iteritems():
        report[code]['imbalance'] = code_imbalance

    # Manual Values not provided by the classification report
    report['kappa'] = cohen_kappa_score(labels_list, predictions_list)
    report['cv'] = {}
    report['cv']['mean'] = np.mean(score)
    report['cv']['std'] = np.std(score)
    report['samples'] = {}
    report['samples']['train'] = len(splits[0])
    report['samples']['test'] = len(splits[-1])
    report['samples']['features'] = len(splits[0].columns)

    self.vprint('Overall: %.3f  Kappa: %.3f' % (report['accuracy'], report['kappa']))
    self.vprint('Weighted F1-Score: %.3f' % (report['weighted avg']['f1-score']))

    true_matrix = confusion_matrix(labels_list, predictions_list, normalize='true')
    pred_matrix = confusion_matrix(labels_list, predictions_list, normalize='pred')

    return columns, report, true_matrix, pred_matrix
```

**Comment by adriangb** [comment]
This is a lot of code. Would you be able to provide a minimal reproducible example where a Keras Model works okay on its own but not when wrapped with SciKeras? Thank you

**Comment by AbelThar** [comment]
### Using same SciKeras model, directly with Keras, is working


```python
# Co-Opted from SciKeras Model Compilation
meta = {
    'classes_': np.array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]), 
    'target_type_': 'multilabel-indicator', 
    'y_dtype_': np.dtype('uint8'), 
    'y_ndim_': 2, 
    'X_dtype_': np.dtype('float32'), 
    'X_shape_': (79997, 96), 
    'n_features_in_': 96, 
    #'target_encoder_': ClassifierLabelEncoder(loss='categorical_crossentropy'), 
    'n_classes_': 12, 
    'n_outputs_': 1, 
    'n_outputs_expected_': 1, 
    #'feature_encoder_': FunctionTransformer()
}

model = fcn_model(128, meta)
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=1)
columns, report, true_matrix, pred_matrix = cl.classification_metrics(model, [X_train, X_test, y_train, y_test], [0, 0])
```
#### Output from the above cell

    meta: {
     ('classes_', array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]))
      ('target_type_', 'multilabel-indicator')
      ('y_dtype_', dtype('uint8'))
      ('y_ndim_', 2)
      ('X_dtype_', dtype('float32'))
      ('X_shape_', (79997, 96))
      ('n_features_in_', 96)
      ('n_classes_', 12)
      ('n_outputs_', 1)
      ('n_outputs_expected_', 1) 
    }
    2500/2500 [==============================] - 3s 1ms/step - loss: 1.2370 - accuracy: 0.5477
    Overall: 0.660  Kappa: 0.615
    Weighted F1-Score: 0.651

Issue is, for SciKeras, the scores above are essentially 0.

**Comment by adriangb** [comment]
Thank you. I will take a look in the next couple of days and try to reproduce.

**Comment by AbelThar** [comment]
Appreciate the help, let me know if I can provide anything.

**Comment by adriangb** [comment]
I think this may be a bug!

Can you test the `fix-multiclass-onehot-inverse` branch I just pushed? https://github.com/adriangb/scikeras/tree/fix-multiclass-onehot-inverse

**Comment by AbelThar** [comment]
Sorry for a bit of a late I have been having some issues building the repository. 

It is not really related to this issue, but maybe you could help me build it - Using conda...:

```yml
name: sci
channels:
  - conda-forge
  - defaults
dependencies:
  - pandas
  - numpy
  - dask
  - pytables
  - pyarrow
  - scikit-learn
  - tensorflow
  - keras
  - ipykernel
  - ipywidgets
  - jupyterlab
  - pip
  - setuptools 
  - poetry
  - pip:
    # - scikeras
    - "--editable=git+https://github.com/adriangb/scikeras.git@fix-multiclass-onehot-inverse#egg=scikeras"
```

While creating this environment it fails because `setuptools` does not find `setup.py`. 
Obviously this is expected, so I attempted to do it manually with `poetry` as documented in [CONTRIBUTING.md](https://github.com/adriangb/scikeras/blob/fix-multiclass-onehot-inverse/CONTRIBUTING.md), but it fails with the following:

```sh
(sci)$ cd src/scikeras/
(sci)$ poetry install
Installing dependencies from lock file
                                  
[NonExistentKey]   
'Key "hashes" does not exist.'  
                                  
install [--no-dev] [--dry-run] [-E|--extras EXTRAS] [--develop DEVELOP]

(sci)$ poetry run pytest tests/
                                                           
[FileNotFoundError]                         
[Errno 2] No such file or directory: 'pytest': 'pytest'  
                                                           
run <args> (<args>)...
```

**Comment by adriangb** [comment]
For simple cases like this, where you just want to test some code from git, but not necessarily push back any changes, the easiest way is to just:

1. Activate your existing environment or a new one (using conda, virtualenv or anything)
2. Use `pip` to install directly from git: `pip pip install git+https://github.com/adriangb/scikeras.git@fix-multiclass-onehot-inverse`

**Comment by AbelThar** [comment]
OK tested, worked as expected. Appreciate the help and the quick response.

**Comment by adriangb** [comment]
Great to hear! Once I merge that PR, that branch will be deleted, so you'll have to work off of master until the next release (at which point you can go back to using the latest public version).

Thank you for reporting this bug and for the detailed logs.

**Comment by AbelThar** [comment]
No problem, it was my pleasure to help, especially with the quick response.
